src/Orleans/IDs/Interner.cs

```diff
@@ -96,31 +96,34 @@ public Interner(int initialSize, TimeSpan cleanupFreq)
         /// <param name="key">key to find</param>
         /// <param name="creatorFunc">function to create new object and store for this key if no cached copy exists</param>
         /// <returns>Object with specified key - either previous cached copy or newly created</returns>
-        public T FindOrCreate(K key, Func<T> creatorFunc)
+        public T FindOrCreate(K key, Func<K, T> creatorFunc)
         {
-            T obj = null;
-            WeakReference cacheEntry = internCache.GetOrAdd(key,
-                (k) =>
-                {
-                    obj = creatorFunc();
-                    return new WeakReference(obj);
-                });
-            if (cacheEntry != null)
+            T result;
+            WeakReference<T> cacheEntry;
+
+            // Attempt to get the existing value from cache.
+            internCache.TryGetValue(key, out cacheEntry);
+
+            // If no cache entry exists, create and insert a new one using the creator function.
+            if (cacheEntry == null)
```

Comment:
What if under `if (cacheEntry == null)` we simply called `cacheEntry = internCache.GetOrAdd()` like before to maintain the behavior in the add case? While the code looks safe to me for how it's used, it's not equivalent to the internal implementation of `GetOrAdd` and can create multiple copies of the cached item. Not a big deal though IMO.

---
In reply to: [91662953](https://github.com/dotnet/orleans/pull/2481#discussion_r91662953) [](ancestors = 91662953)